File: aoc2023/aoc6.py
#!/usr/bin/env python

from aoc_util import get_data_lines, numbers
import re

data = get_data_lines(6)
times = numbers(data[0])
dists = numbers(data[1])
total = 1
for t, d in zip(times, dists):
    myways = 0
    for h in range(t + 1):
        if h * (t - h) > d:
            myways += 1
    total *= myways
print(total)

(bigtime,) = numbers(data[0].replace(" ", ""))
(bigdist,) = numbers(data[1].replace(" ", ""))
minh1 = 0
minh2 = bigtime // 2 - 1
while minh2 - minh1 > 1:
    mid = (minh1 + minh2) // 2
    if mid * (bigtime - mid) > bigdist:
        minh2 = mid
    else:
        minh1 = mid
maxh1 = bigtime // 2 + 1
maxh2 = bigtime
while maxh2 - maxh1 > 1:
    mid = (maxh1 + maxh2) // 2
    if mid * (bigtime - mid) > bigdist:
        maxh1 = mid
    else:
        maxh2 = mid
# print(minh1,minh2,maxh1,maxh2)
print(maxh2 - minh2)

File: aoc2023/aoc6.hs
import System.Environment (getArgs)
import Data.Char (isDigit)

findHoldMin :: Int -> Int -> Int
findHoldMin time dist = snd (res (0, time `div` 2 - 1))
    where
        res (lo, hi) | hi - lo <= 1 = (lo, hi)
        res (lo, hi) =
                let mid = (lo + hi) `div` 2
                 in if mid * (time - mid) > dist then res (lo, mid) else res (mid, hi)

findHoldMax :: Int -> Int -> Int
findHoldMax time dist = fst (res (time `div` 2 - 1, time))
    where
        res (lo, hi) | hi - lo <= 1 = (lo, hi)
        res (lo, hi) =
                let mid = (lo + hi) `div` 2
                 in if mid * (time - mid) > dist then res (mid, hi) else res (lo, mid)

findHoldPoss :: Int -> Int -> Int
findHoldPoss time dist = 1 + findHoldMax time dist - findHoldMin time dist

main :: IO ()
main = do
        args <- getArgs
        let filename = if null args then "aoc6.in" else head args
        s <- lines <$> readFile filename
        let times = read <$> tail (words $ head s)
        let dists = read <$> tail (words $ head (tail s))
        let p1holdtimes = uncurry findHoldPoss <$> zip times dists
        print $ product p1holdtimes
        let bigtime = read $ filter isDigit (head s)
        let bigdist = read $ filter isDigit (head $ tail s)
        print $ findHoldPoss bigtime bigdist
